### DISCUSSION TITLE
How do I get all the mapped paths of the system Routes handlerDef using java?

### DISCUSSION THREAD

**Discussion by coolwasp** [playframework/playframework#11873] [Q&A]
If it is available, it can be processed in advance, such as router rule processing or saving rules to a database for authorization，Or you can automatically generate API documents.

**Comment by tsuyoshizawa** [comment]
By getting `RoutesProvider` in Play Framework 2.8, it is possible to get the HTTP methods, paths, and Controller method names listed in the routes file. This is an example for Scala, but the same is true for Java.

```scala
class HomeController @Inject()(routesProvider: RoutesProvider) extends BaseController {
  def index(): Action[AnyContent] = Action {
     routesProvider.get.documentation.foreach { case (httpMethod, pathPattern, controllerMethodInvocation) =>
      println(httpMethod, pathPattern, controllerMethodInvocation)
    }

    Ok(views.html.index())
  }
}
```

Result:

```
(GET,/,controllers.HomeController.index())
(GET,/assets/$file<.+>,controllers.Assets.versioned(path:String = "/public", file:Asset))
```

**Reply by coolwasp** [reply]
Thank you very much for your reply. The method you provided is also very good, and you can obtain basic information about all routes. It would be even better if you could obtain HandlerDef information for all routes.

package play.api.routing

/**
 * Information about a `Handler`, especially useful for loading the handler
 * with reflection.
 */
case class HandlerDef(
    classLoader: ClassLoader,
    routerPackage: String,
    controller: String,
    method: String,
    parameterTypes: Seq[Class[_]],
    verb: String,
    path: String,
    comments: String = "",
    modifiers: Seq[String] = Seq.empty
) extends play.routing.HandlerDef
